os/20211014-s390x-secure-execution.md

```diff
@@ -0,0 +1,171 @@
+  `-crypt_rootfs root         crypt /sysroot
+
+```
+
+- `se`   - This is a new partition `ext4`, where we store a `sd-boot` encrypted image
+- `boot` - boot partition is the same as one on reqular qcow2 image, additionally it's encrypted with random LUKS key during CoreOS build process
+- `root` - rootfs partition is almost the same as one on reqular qcow2 image, additionally it's encrypted with random LUKS key during CoreOS build process
+and contains `/etc/crypttab`, `/etc/luks/root` and `/etc/luks/boot` files.
+
+The reason why `boot` has to be encrypted is that without encryption hacker is able to mount it and modify its contents, for example add new kargs to BLS config file(s), so during system's lifetime that modification could affect `sd-boot` image.
+
+## Firstboot
+
+During guest provisioning new hostkey(s) is/are fetched and used for generation of `sd-boot`. User should specify them within ignition config with pre-defined name template: `/etc/se-hostkeys/ibm-z-hostkey-N` where `N` is a key number, several hostkeys could be used:
+
+```yaml
+storage:
+  files:
+    -
+      path: /etc/se-hostkeys/ibm-z-hostkey-1
+      overwrite: true
+      contents:
+        source: https://url.com/key1
+```
+
+System would also set a new random LUKS keys for both `boot` and `root` filesystems, so each node has its own LUKS keys.
+
+## Lifetime
+
+SecureExecution is enabled by 2 tools provided by IBM's s390-utils package:
+- `genprotimg` - tool to generate `sd-boot`
+- `zipl` - tool to install new bootloader
+
+Despite `zipl` is well known and used by CoreOS, it couldn't be executed same way as before. When system runs in SecureExecution, `ostree` and `rdcore` automatically detect that and perform all required steps.
+That means, that user won't notice any difference, both mentioned components would automatically regenerate `sd-boot` image and
+install it with `zipl` whenever required.
+
+# Implementation steps
+
+## Step 1: CoreOS with existing patches to `libostree` could be configured to perform in SecureExecution mode.
+
+Customer public hostkeys should be placed into `/etc/se-hostkeys/` using name templates `ibm-z-hostkey-N`. SecureExecution is enabled by hands following https://www.ibm.com/docs/en/linux-on-systems?topic=linux-introduction.
+
+## Step 2: CoreOS with is always performing in SecureExecution mode.
+
+Pre-build SecureExecution-ready qcow2 with LUKS encrypted `boot-` and `rootfs` is available for download. The image contains the `sd-boot` image signed with universal public hostkey, which is valid for all Z15 and newer machines. Customer should add his public hostkeys(s) into `ignition` config.
+
+## Step 3: Reencryption of `boot-` and `rootfs` with random LUKS keys and `integrity` option.
+
+During firstboot, the LUKS keys will be changed to a randomly generated ones. The idea behind this is to set on each host the LUKS keys to a unique keys during firstboot. This should prevent reading the key from the provided pre-build image.
```

Comment:
> The pristine image will also need to ship with integrity protection enabled.

This doesn't support `growfs`

Reply:
Then we'll need to reprovision the root filesystem rather than using `growfs`.  If we don't enable integrity protection in the pristine image, a malicious host can modify the disk image before it's booted.

Reply:
yes, during LUKS keys modification we'd have to do this